### PR TITLE
[WIP]API: CategoricalType for specifying categoricals

### LINKED ISSUE #14676: API: Expand DataFrame.astype to allow Categorical(categories, ordered)
#### A small, complete example of the issue

This is a proposal to allow something like

```python
df.astype({'A': pd.CategoricalDtype(['a', 'b', 'c', 'd'], ordered=True})
```

Currently, it can be awkward to convert many columns in a DataFrame to a Categorical *with control over the categories and orderedness*. If you just want to use the defaults, it's not so bad with `.astype`:

```python
In [5]: df = pd.DataFrame({"A": list('abc'), 'B': list('def')})

In [6]: df
Out[6]:
   A  B
0  a  d
1  b  e
2  c  f

In [8]: df.astype({"A": 'category', 'B': 'category'}).dtypes
Out[8]:
A    category
B    category
dtype: object
```

If you need to control `categories` or `ordered`, your best off with

```python
In [20]: mapping = {'A': lambda x: x.A.astype('category').cat.set_categories(['a', 'b'], ordered=True),
    ...:            'B': lambda x: x.B.astype('category').cat.set_categories(['d', 'f', 'e'], ordered=False)}

In [21]: df.assign(**mapping)
Out[21]:
     A  B
0    a  d
1    b  e
2  NaN  f
```

By expanding astype to accept instances of `Categorical`, you remove the need for the `lambda`s and you can do conversions of other types at the same time.

This would mirror the semantics in https://github.com/pandas-dev/pandas/issues/14503

Updated to change `pd.Categorical(...)` to a new/modified `pd.CategoricalDtype(...)` based on the discussion below.

### PR DESCRIPTION
**Pull request by TomAugspurger**
closes #14676 

This adds a new top-level class `pd.CategoricalType` for representing the *type* information on a `Categorical`, *without the values*. Users can pass this anywhere they would previously pass `'category'` if they need to specify non-default values for `categories` or `ordered` (you can still use `'category'` of course):

```python
In [1]: from pandas import *

In [2]: paste
    >>> t = CategoricalType(categories=['b', 'a'], ordered=True)
    >>> s = Series(['a', 'a', 'b', 'b', 'a'])
    >>> c = s.astype(t)
    >>> c

## -- End pasted text --
Out[2]:
0    a
1    a
2    b
3    b
4    a
dtype: category
Categories (2, object): [b < a]
```

This is the simplest possible change for now. A bigger change would be to make `c.dtype` return a `CategoricalType` instead of '`category`'. We could probably do that in a way that's backwards compatible, but I'll need to think on it a bit.

Other places this should work

- `select_dtypes` (currently returns all categoricals; should just be ones with same `CategoricalType`?)
- `read_csv`: need to implement, see https://github.com/pandas-dev/pandas/issues/14503
- `Series(..., dtype=...`)
- probably others.

Implementation-wise I need to document, more tests, clean up some things like the repr. But I wanted to get this out there for discussion. @janschulz you might be interested.

### CODE REVIEW

File: pandas/core/categorical.py [hunk 1]
```diff
@@ -2066,3 +2066,48 @@ def _factorize_from_iterables(iterables):
         # For consistency, it should return a list of 2 lists.
         return [[], []]
     return map(list, lzip(*[_factorize_from_iterable(it) for it in iterables]))
+
+
+class CategoricalType(CategoricalDtype):
+    """
```
Comment (jreback) [inline pandas/core/categorical.py]: this should be in with the rest of the dtypes

to be honest i wouldn't create this; just add it optionally into the existing

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/categorical.py (hunk 1 above)
Comment (TomAugspurger) [inline pandas/core/categorical.py]: Is that OK with the caching and stuff that is done on `CategoricalDType`? This is on `CategoricalyDtype`:

```python
    def __new__(cls):

        try:
            return cls._cache[cls.name]
        except KeyError:
            c = object.__new__(cls)
            cls._cache[cls.name] = c
            return c
```

I haven't messed with extension types much. We could make the keys of that internal dict reflect the categories and ordered attributes.

Comment (jreback) [inline pandas/core/categorical.py]: yes you would had optional attributes and cache based on them

Comment (TomAugspurger) [inline pandas/core/categorical.py]: K, this seems to be working. Just have to convert the categories to a tuple so that they're hashable. Thanks.

**Comment by TomAugspurger** [comment]
What would the ideal `__repr__` of `CategoricalDtype` be? Currently it's just `category`. For places like `DataFrame.info` it's probably best to keep that as is, but that's not as useful if you're inpsecing `CategoricalDtype` itself.

Also, for equality semantics, this is what I have right now

```python
    An instance of ``CategoricalDtype`` compares equal with any other
    instance of ``CategoricalDtype``, regardless of categories or ordered.
    In addition they compare equal to the string ``'category'``.
    To check whether two instances of a ``CategoricalDtype`` match,
    use the ``is`` operator.

    >>> t1 = CategoricalDtype(['a', 'b'], ordered=True)
    >>> t2 = CategoricalDtype(['a', 'c'], ordered=False)
    >>> t1 == t2
    True
    >>> t1 == 'category'
    True
    >>> t1 is t2
    False
    >>> t1 is CategoricalDtype(['a', 'b'], ordered=True)
    True
```

though I don't expect people to be working with these objects that much.

File: pandas/types/dtypes.py [hunk 2]
```diff
@@ -98,25 +99,73 @@ class CategoricalDtypeType(type):
 class CategoricalDtype(ExtensionDtype):
 
     """
+    Type for categorical data with the categories and orderedness,
+    but not the values
+
+    .. versionadded:: 0.20.0
```
Comment (jreback) [inline pandas/types/dtypes.py]: this has been around for quite some time, you are adding parameter support in 0.20.0

File: pandas/types/dtypes.py [hunk 3]
```diff
@@ -98,25 +99,73 @@ class CategoricalDtypeType(type):
...
+
+    Parameters
+    ----------
+    categories : list or None
```
Comment (jreback) [inline pandas/types/dtypes.py]: list-like, use similar to whats in Categorical now

File: pandas/types/dtypes.py [hunk 4]
```diff
@@ -98,25 +99,73 @@ class CategoricalDtypeType(type):
...
+    ordered : bool, default False
+
+    Examples
+    --------
+    >>> t = CategoricalDtype(categories=['b', 'a'], ordered=True)
```
Comment (jreback) [inline pandas/types/dtypes.py]: these examples are not relevant. This should be not include Series. This is a self-contained type.

File: pandas/types/dtypes.py [hunk 5]
```diff
@@ -98,25 +99,73 @@ class CategoricalDtypeType(type):
...
     name = 'category'
     type = CategoricalDtypeType
     kind = 'O'
     str = '|O08'
     base = np.dtype('O')
     _cache = {}
 
-    def __new__(cls):
+    def __new__(cls, categories=None, ordered=False):
+        categories_ = categories if categories is None else tuple(categories)
```
Comment (jreback) [inline pandas/types/dtypes.py]: this needs all of the validation logic (from Categorical). on the actual categories.

**Comment by jreback** [comment]
you need to move over the repr (of the categories) and the validation logic on the categories to this class.

Further Categorical should use this internally for storage of categories / ordered. 

This is a fairly invasive change and needs some thought.

**Comment by jreback** [comment]
I would split off the actual issue change and make that a follow up PR. Just putting in place the correct infrastructure can be the scope of this PR>

**Comment by TomAugspurger** [comment]
> This is a fairly invasive change and needs some thought.

Yep. I had hoped to do the minimal change of just providing a new API, but let's do it right the first time. I'll spend some time on this the next couple weeks and ping when I have something.

**Comment by TomAugspurger** [comment]
Closing for now. Will reopen a new PR on top of a PR fixing https://github.com/pandas-dev/pandas/issues/14711

**Pull request closed without merging**